app/src/main/java/com/danosoftware/galaxyforce/services/googleplay/GooglePlayServices.java

```diff
@@ -0,0 +1,225 @@
+import com.google.android.gms.auth.api.signin.GoogleSignInOptions;
+import com.google.android.gms.common.api.ApiException;
+import com.google.android.gms.drive.Drive;
+import com.google.android.gms.games.Games;
+import com.google.android.gms.games.GamesClient;
+import com.google.android.gms.games.SnapshotsClient;
+import com.google.android.gms.tasks.OnCompleteListener;
+import com.google.android.gms.tasks.Task;
+
+import static com.danosoftware.galaxyforce.constants.GameConstants.RC_SIGN_IN;
+
+public class GooglePlayServices {
+
+    /* logger tag */
+    private static final String ACTIVITY_TAG = "GooglePlayServices";
+
+    private enum ConnectedState {
+        NO_ATTEMPT, CONNECTED, DISCONNECTED;
+    }
+
+    private final Activity mActivity;
+    private final GoogleSignInClient signInClient;
+    private final GoogleSignInOptions signInOptions;
+    private ConnectedState connectedState;
+
+    public GooglePlayServices(Activity activity) {
+        Log.d(ACTIVITY_TAG, "Creating Billing client.");
+        this.mActivity = activity;
+        this.signInOptions =
+                new GoogleSignInOptions.Builder(GoogleSignInOptions.DEFAULT_GAMES_SIGN_IN)
+                        // Add the APPFOLDER scope for Snapshot support.
+                        .requestScopes(Drive.SCOPE_APPFOLDER)
+                        // below scopes are replace the above deprecated scope
+//                        .requestScopes(new Scope(DriveScopes.DRIVE_APPDATA))
+//                        .requestScopes(new Scope(DriveScopes.DRIVE_FILE))
+                        .build();
+        this.signInClient = GoogleSignIn.getClient(
+                activity,
+                signInOptions);
+        this.connectedState = ConnectedState.NO_ATTEMPT;
+    }
+
+    /**
+     * Attempt to sign-in without interupting the user.
+     * If previous attempts have resulted in us being disconected,
+     * do not try again.
+     */
+    public void signInSilently() {
+        GoogleSignInAccount account = GoogleSignIn.getLastSignedInAccount(mActivity);
+        if (GoogleSignIn.hasPermissions(account, signInOptions.getScopeArray())) {
+            // Already signed in.
+            onConnected(account);
+        } else {
+            // Not signed-in. Try the silent sign-in first.
+            signInClient.silentSignIn().addOnCompleteListener(mActivity,
+                    new OnCompleteListener<GoogleSignInAccount>() {
+                        @Override
+                        public void onComplete(@NonNull Task<GoogleSignInAccount> task) {
+                            if (task.isSuccessful()) {
+                                Log.i(ACTIVITY_TAG, "Success signInSilently");
+                                GoogleSignInAccount signedInAccount = task.getResult();
+                                onConnected(signedInAccount);
+                            } else {
+                                Log.w(ACTIVITY_TAG, "Failed signInSilently");
```

Comment:
At this point, we can call `getException()` to obtain an `ApiException` with the detailed status code. A status code of `CommonStatusCodes.SIGN_IN_REQUIRED` indicates that the player needs to take explicit action to sign-in.  See: https://developers.google.com/games/services/android/signin